archive/100-theorems-list/pnat_perfect.lean

```diff
@@ -0,0 +1,509 @@
+  { rw k_ih, conv_rhs {rw geom_series_succ}, rw add_comm, simp },
+  { intro contra,
+    have h := divisor_le contra,
+    have g1 :=  nat.mul_lt_mul_of_pos_left (nat.prime.one_lt pp) (nat.pos_pow_of_pos (k_n) (nat.prime.pos pp)),
+    apply not_lt_of_le h, rw pow_add, rw ← pnat.coe_lt_coe, revert g1, simp }
+end
+
+lemma sum_divisors_prime {p : ℕ+} (pp : p.prime) : sum_divisors p = p + 1 :=
+begin
+  have h := sum_divisors_pow_prime pp 1, rw pow_one at h, rw h, unfold geom_series,
+  simp [finset.sum_range_succ]
+end
+
+lemma pnat.prime_two : (2 : ℕ+).prime := nat.prime_two
+
+lemma geom_series_two (k : ℕ): geom_series 2 k = 2 ^ k - 1 :=
+begin
+  induction k with k hk, simp,
+  rw [geom_series_succ, hk, nat.pow_succ, mul_two, add_comm, nat.add_sub_assoc, nat.pow_eq_pow],
+  apply nat.pos_pow_of_pos, omega
+end
+
+lemma sum_divisors_pow_two (k : ℕ) : sum_divisors (2 ^ k) = (2 ^ (k + 1) - 1) :=
+begin
+  induction k with k hk, simp [sum_divisors],
+  rw sum_divisors_pow_prime pnat.prime_two k.succ, simp [geom_series_two]
+end
+
+lemma odd_mersenne_succ (k : ℕ) : ¬  2 ∣ (2 ^ (k + 1) - 1) :=
+begin
+  intro contra,
+  have h : 2 ∣ 2 ^ (k + 1) - 1 + 1,
+  { existsi 2 ^ k,
+    rw [← nat.pred_eq_sub_one, ← nat.succ_eq_add_one, nat.succ_pred_eq_of_pos _],
+    { rw [mul_comm, nat.pow_succ] },
+    apply nat.pos_pow_of_pos, omega },
+  rw ← nat.dvd_add_iff_right contra at h,
+  have h:= nat.le_of_dvd _ h; linarith,
+end
+
+lemma odd_sum_divisors_pow_two (k : ℕ) : ¬  2 ∣ (sum_divisors (2 ^ k)) :=
+begin
+  rw sum_divisors_pow_two k, apply odd_mersenne_succ,
+end
+
+@[simp]
+lemma pnat.coe_eq_one_iff {m : ℕ+} :
+(m : ℕ) = 1 ↔ m = 1 := by { split; intro h; try { apply pnat.eq}; rw h; simp }
+
+
+lemma pnat.dvd_prime {p m : ℕ+} (pp : p.prime) :
+(m ∣ p ↔ m = 1 ∨ m = p) := by { rw pnat.dvd_iff, rw nat.dvd_prime pp, simp }
+
+lemma divisors_prime {p : ℕ+} (pp : p.prime) : divisors p = {p, 1} :=
+begin
+  ext,
+  simp only [mem_divisors, ne.def, finset.mem_insert, finset.mem_singleton],
+  split; intro h, rw pnat.dvd_prime pp at h, apply h.symm,
+  cases h; rw h, apply pnat.dvd_refl, apply pnat.one_dvd
+end
+
+lemma card_pair_eq_two {α : Type} {x y : α} (neq : x ≠ y) : ({x, y} : finset α).card = 2 :=
+begin
+  rw finset.card_insert_of_not_mem, rw finset.card_singleton, rw finset.mem_singleton, apply neq,
+end
+
+----why can't I use the above????
+lemma pnat.card_pair_eq_two {x y : ℕ+} (neq : x ≠ y) : ({x, y} : finset ℕ+).card = 2 :=
+begin
+  rw finset.card_insert_of_not_mem, rw finset.card_singleton, rw finset.mem_singleton, apply neq,
+end
+
+lemma pnat.prime.ne_one {p : ℕ+} : p.prime → p ≠ 1 :=
+by { intro pp, intro contra, apply nat.prime.ne_one pp, rw pnat.coe_eq_one_iff, apply contra }
+
+lemma pnat.eq_one_of_lt_two {n : ℕ+} : n < 2 → n = 1 :=
+begin
+  intro h, apply le_antisymm, swap, apply pnat.one_le,
+  change n < 1 + 1 at h, rw pnat.lt_add_one_iff at h, apply h
+end
+
+@[simp]
+lemma pnat.one_val : (1 : ℕ+).val = 1 := rfl
```

Comment:
This lemma is completely fine, but the `.val` shouldn't show up in your context. If it does, it means that some lemma is using `.val` instead of the coercion, and we should fix that.